docs/source/model_design/programming_vivarium/debugging/index.rst

```diff
@@ -0,0 +1,190 @@
+significantly removed from the line of code that will need to be changed 
+to fix the error. 
+
+Let's work through an example to understand this. Say a line of code 
+we edit in the :code:`loader.py` file filters a dataset incorrectly 
+such that it is actually empty. Nothing in that immediate block of 
+code catches this error - instead the provided dataframe is just 
+stored as empty. 
+
+From there this same, empty dataset is used in another :code:`loader.py` 
+function to calculate a rate. So our erroreously empty dataframe will be 
+multiplied by a dataframe with the correct data in it. 
+When pandas tries to by this multiplication on the dataframes, 
+it will find that the lengths do not match when they should. Pandas has 
+functions that run through a series of checks and if one fails, throws an 
+error message. 
+
+So the bottom line of the stack trace will likely say an error was thrown 
+in a pandas error message function that the lengths are not equal. From there, 
+the stack trace displays the code that lead here - for example the pandas 
+checking function that was run, then the pandas function that was called 
+(the multiplication), then the secondary 
+:code:`loader.py` function that tried to multiply with the empty dataframe. And still, 
+none of these will actually reveal the issue with filtering the dataset 
+we started with! This can be further complicated if the :code:`loader.py` function 
+using some functions for other packages, like :code:`vivarium_research` or :code:`vivarium_public_health`. 
+
+As you can see from this example, the stack trace can be both helpful and unhelpful. 
+In general, look through the files to see if you can get back to something 
+you wrote or edited - like the :code:`loader.py` file. If so, this is a good 
+place to start looking for issues, and the stack trace will provide the line 
+number of the code that caused the error. 
+
+Otherwise, see the next section for how to use the python debugger. 
+
+Navigating the Python Debugger
+------------------------------
+
+In all of the commands used to run engineering commands, the 
+flag :code:`--pbd` is used, which means that when an error 
+occurs, you will be placed into the python debugger. 
+
+The python debugger is nice to use in tandem with the stack trace. 
+Let's continue with the example above, where we got an error 
+in a pandas checking function. 
+
+The python debugger places you at the end of the stack trace - 
+so we are currently sitting in a pandas error message function. 
+We can check this by typing :code:`l.` or :code:`ll` which prints 
+the code around the current line (l.), or a longer version of this 
+print out (ll). This will let you see what the code surrounding the 
+error is. Likely in this case, it will just show you that pandas found 
+mismatched lengths and threw the error message, which isn't too helpful. 
+
+Next, you can move up or down the stack trace with :code:`u` or 
```

Comment:
I think when you say "stack trace" here, it should be "call stack." The "trace" is just how that stack gets output in e.g. an error message. It could also be good to explain what the "stack" is: it's the code you're currently running, and what function called the current one, and what function called that one, and so on, all the way up the top.

Reply:
Thanks!